src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs::DirBuilder;
+use std::io::{Error, ErrorKind, Result};
+use std::os::unix::fs::DirBuilderExt;
+use std::path::{Path, PathBuf};
+
+use crate::{safe_join, SafePathBuf};
+
+const DIRECTORY_MODE_DEFAULT: u32 = 0o700;
+const DIRECTORY_MODE_MASK: u32 = 0o777;
+
+/// Safe version of `DirBuilder` to protect from TOCTOU style of attacks.
+#[derive(Debug)]
+pub struct SafeDirBuilder {
```

Comment:
As with `SafePathBuf`, I don't think "safe" is a good name because it doesn't specify what particular properties are safe about it.  In this case I think what you're trying to guarantee is that it will never create any directories outside `root`.  So, maybe `ScopedDirBuilder` or `RootedDirBuilder`?

Reply:
Really poor at naming. Actually the `SafeDirBuilder` has three properties:
```
/// The `SafeDirBuilder` is a counterpart for `DirBuilder`, with safety enhancements of:
/// - ensuring the new directories are created under a specified `root` directory.
/// - avoiding symlink race attacks during making directories
/// - returning a [PinnedPathBuf] for the last level of directory, so it could be used for other
///   operations safely.
```